File: Java/src/main/java/QuestionBank/Solution101_200/Solution130.java
package QuestionBank.Solution101_200;

/**
 * @author windows
 */
public class Solution130 {
    public void solve(char[][] board) {
        int m = board.length;
        int n = board[0].length;
        for (int i = 0; i < m; i++) {
            dfs(board, i , 0);
            dfs(board, i , n - 1);
        }
        for (int i = 0; i < n; i++) {
            dfs(board, 0 , i);
            dfs(board, m - 1 , i);
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 'A'){
                    board[i][j] = 'O';
                }else if (board[i][j] == 'O'){
                    board[i][j] = 'X';
                }
            }
        }
    }

    private void dfs(char[][] board, int x, int y){
        if (x < 0 || y < 0 || x >= board.length || y >= board[0].length || board[x][y] != 'O'){
            return;
        }
        board[x][y] = 'A';
        dfs(board, x + 1, y);
        dfs(board, x - 1, y);
        dfs(board, x, y + 1);
        dfs(board, x, y - 1);
    }

    public static void main(String[] args) {
        Solution130 solution130 = new Solution130();
        solution130.solve(new char[][]{{'X','O','X','O','X','O'},{'O','X','O','X','O','X'},{'X','O','X','O','X','O'},{'O','X','O','X','O','X'}});
    }
}

File: C++/src/Solution101_200/Solution130.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution130.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include "vector"
using namespace std;

void dfs(vector<vector<char>>& board, int i, int j){
  if (i <= 0 || j <= 0 ||i >= board.size() || j >= board[0].size() || board[i][j] != 'O'){
    return;
  }
  board[i][j] = 'Z';
  dfs(board, i - 1, j);
  dfs(board, i + 1, j);
  dfs(board, i, j - 1);
  dfs(board, i, j + 1);
}


void solve(vector<vector<char>>& board) {
  for (int i = 0; i < board[0].size(); ++i) {
    dfs(board, 0, i);
    dfs(board, board.size() - 1, i);
  }
  for (int i = 0; i < board.size(); ++i) {
    dfs(board, i, 0);
    dfs(board, i, board[0].size() - 1);
  }
  for (int i = 0; i < board.size(); ++i) {
    for (int j = 0; j < board[0].size(); ++j) {
      if (board[i][j] == 'O'){
        board[i][j] = 'X';
      }
      if (board[i][j] == 'Z'){
        board[i][j] = 'O';
      }
    }
  }
  return;
}
